### PR TITLE
HCL2: add templatefile function

### PR DESCRIPTION
**Pull request by azr**
helps with #9485

### CODE REVIEW

File: website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+sequences and directives.
+
+Instead, you can write a template that consists only of a single interpolated
+call to either jsonencode or yamlencode, specifying the value to encode using
+normal expression syntax as in the following examples:
+
+```hcl
+${jsonencode({
+  "backends": [for addr in ip_addrs : "${addr}:${port}"],
+})}
```
Comment (sylviamoss) [inline website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx:95]: This is nice!

File: website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+```hcl
+%{ for addr in ip_addrs ~}
+backend ${addr}:${port}
+%{ endfor ~}
+```
+The templatefile function renders the template:
+
+```shell-session
+> templatefile("${path.root}/backends.tpl", { port = 8080, ip_addrs = ["10.0.0.1", "10.0.0.2"] })
```
Comment (sylviamoss) [inline website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx:51]:
```suggestion
> templatefile("${path.root}/backends.tmpl", { port = 8080, ip_addrs = ["10.0.0.1", "10.0.0.2"] })
```

Would any extension work?

**Review by sylviamoss — APPROVED** [review]
This LGTM 👍🏼

File: website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx (hunk 1 above)
Comment (azr) [reply website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx:95]: Yes 🙂  !

File: website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  The templatefile function reads the file at the given path and renders its
+  content as a template using a supplied set of template variables.
+---
+
+# `templatefile` Function
+
+-> *Recommendation:* we recommend on using the `.pkrtpl.hcl` extension file,
+when using the `templatefile`, these files *are* hcl files but also templates
+and a slightly different set of features than the ones offered in a Packer
+template, whilst still being specific to Packer.
```
Comment (azr) [inline website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx]: Making the recommended file extension `.pkrtpl.hcl` ! I hope everyone agrees !

File: website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  The templatefile function reads the file at the given path and renders its
+  content as a template using a supplied set of template variables.
+---
+
+# `templatefile` Function
+
+-> *Recommendation:* we recommend on using the `.pkrtpl.hcl` extension file,
+when using the `templatefile` functions, template files *are* hcl files but also
+templates and have slightly different set of features than the ones offered in a
+Packer template, whilst still being specific to Packer.
```
Comment (azr) [inline website/content/docs/templates/hcl_templates/functions/file/templatefile.mdx:14]: Making the recommended file extension `.pkrtpl.hcl` ! I hope everyone agrees !

**Comment by SwampDragons** [comment]
I think this is a complex enough concept that it would benefit from a real example using http_content, this function, and a real preseed example using some templated custom IP address.

**Comment by SwampDragons** [comment]
Code looks good. Thanks for the tests.

**Comment by azr** [comment]
> I think this is a complex enough concept that it would benefit from a real example using http_content, this function, and a real preseed example using some templated custom IP address.

I totally agree, I will open a new PR once this one and the #10801 are merged.

**Pull request merged**